### PR TITLE
ramips: add support for ZyXEL Keenetic Extra II

### PR DESCRIPTION
**Pull request by Linaro1985**
Specification:
- System-On-Chip: MT7628N/N
- CPU/Speed: 580 MHz
- Flash-Chip: Winbond w25q256
- Flash size: 32768 KiB
- RAM: 128 MiB
- 5x 10/100 Mbps Ethernet
- 4x external, non-detachable antennas
- UART (J1) header on PCB (57600 8n1)
- Wireless No1 (2T2R): SoC-integrated: MT7628N 2.4GHz 802.11bgn
- Wireless No2 (2T2R): On-board chip: MT7612EN 5GHz 802.11ac
- USB: Yes 1 x 2.0
- 4x LED, 3x button

The device supports dual boot mode. So we use only first half of flash.

Flash instruction:

The only way to flash OpenWrt image is to use
tftp recovery mode in U-Boot:

1. Configure PC with static IP 192.168.1.2/24 and tftp server.
2. Rename "openwrt-ramips-mt76x8-zyxel_keenetic-extra-ii-squashfs-factory.bin"
   to "kextra2_recovery.bin" and place it in tftp server directory.
3. Connect PC with one of LAN ports, press the reset button, power up
   the router and keep button pressed until power led start blinking.
4. Router will download file from server, write it to flash and reboot.

### CODE REVIEW

File: target/linux/ramips/image/mt76x8.mk [hunk 1]
```diff
@@ -57,6 +61,19 @@ define Device/hc5661a
 endef
 TARGET_DEVICES += hc5661a
 
+define Device/ki_rb
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: `define Device/zyxel_keenetic-extra-ii` please

File: target/linux/ramips/image/mt76x8.mk [hunk 2]
```diff
@@ -57,6 +61,19 @@ define Device/hc5661a
...
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  UIMAGE_NAME := ZyXEL Keenetic Extra II
+  DEVICE_TITLE := ZyXEL Keenetic Extra II
+  SUPPORTED_DEVICES := zyxel,keenetic-extra-ii
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport
+  IMAGES += factory.bin
+  IMAGE/factory.bin := $$(IMAGE/sysupgrade.bin) | pad-to 64k | check-size $$$$(IMAGE_SIZE) | \
+        zyimage -d 6162 -v "ZyXEL Keenetic Extra II"
+endef
+TARGET_DEVICES += ki_rb
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: `TARGET_DEVICES += zyxel_keenetic-extra-ii` please

File: target/linux/ramips/image/mt76x8.mk [hunk 3]
```diff
@@ -57,6 +61,19 @@ define Device/hc5661a
 endef
 TARGET_DEVICES += hc5661a
 
+define Device/ki_rb
+  DTS := ki_rb
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  UIMAGE_NAME := ZyXEL Keenetic Extra II
+  DEVICE_TITLE := ZyXEL Keenetic Extra II
+  SUPPORTED_DEVICES := zyxel,keenetic-extra-ii
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: Due to the correct device define, the `SUPPORTED_DEVICES` can be dropped.

File: target/linux/ramips/image/mt76x8.mk [hunk 4]
```diff
@@ -57,6 +61,19 @@ define Device/hc5661a
 endef
 TARGET_DEVICES += hc5661a
 
+define Device/ki_rb
+  DTS := ki_rb
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  UIMAGE_NAME := ZyXEL Keenetic Extra II
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: I'm quite sure this line can be dropped.

File: target/linux/ramips/image/mt76x8.mk [hunk 5]
```diff
@@ -2,6 +2,10 @@
 # MT76x8 Profiles
 #
 
+define Build/zyimage
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: Please move this define to target/linux/ramips/image/Makefile and drop it from target/linux/ramips/image/mt7620.mk as well.

File: target/linux/ramips/image/mt76x8.mk [hunk 6]
```diff
@@ -57,6 +61,19 @@ define Device/hc5661a
 endef
 TARGET_DEVICES += hc5661a
 
+define Device/ki_rb
+  DTS := ki_rb
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: $(ralink_default_fw_size_16M) doesn't match the size of your firmware partition. ralink_default_fw_size_16M == 16121856 bytes == 0xf60000 bytes. Please use 16064k.

File: target/linux/ramips/dts/ki_rb.dts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+			label = "factory";
+			reg = <0x40000 0x10000>;
+			read-only;
+		};
+
+		partition@50000 {
+			label = "firmware";
+			reg = <0x50000 0xfb0000>;
+		};
+	};
```
Comment (mkresin) [inline target/linux/ramips/dts/ki_rb.dts]: would you please add the omitted dual boot related partitions anyway?

**Review by mkresin — CHANGES_REQUESTED** [review]
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

File: target/linux/ramips/image/mt76x8.mk (hunk 1 above)
Comment (Linaro1985) [reply target/linux/ramips/image/mt76x8.mk]: ok

File: target/linux/ramips/image/mt76x8.mk (hunk 3 above)
Comment (Linaro1985) [reply target/linux/ramips/image/mt76x8.mk]: ok

File: target/linux/ramips/image/mt76x8.mk (hunk 4 above)
Comment (Linaro1985) [reply target/linux/ramips/image/mt76x8.mk]: ok

File: target/linux/ramips/image/mt76x8.mk (hunk 5 above)
Comment (Linaro1985) [reply target/linux/ramips/image/mt76x8.mk]: I made this in separate commit.

File: target/linux/ramips/image/mt76x8.mk (hunk 6 above)
Comment (Linaro1985) [reply target/linux/ramips/image/mt76x8.mk]: Firmware partition size is 0xe90000 bytes. So now IMAGE_SIZE is 14912k. It's right?

File: target/linux/ramips/dts/ki_rb.dts (hunk 7 above)
Comment (Linaro1985) [reply target/linux/ramips/dts/ki_rb.dts]: Ok. "u-state" partition store information about which firmware image must be loading.

**Comment by Linaro1985** [comment]
I made changes!
Thanks!

**Pull request closed without merging**